File: leetcode/python/5_longest_palindromic_substring.py
class Solution:
    def longestPalindrome(self, s: str) -> str:
        n = len(s)
        res = ''
        dp = [[False] * n for _ in range(n)]
        for i in range(n):
            dp[i][i] = True
            res = s[i]
        
        for i in range(n-1, -1, -1):
            for j in range(i+1, n):
                if s[i] == s[j]:
                    if j-i == 1 or dp[i+1][j-1]:
                        dp[i][j] = True
                        if len(res) < j - i + 1:
                            res = s[i:j+1]
        return res

s = Solution()
print(s.longestPalindrome("aa"))
print(s.longestPalindrome("abba"))
print(s.longestPalindrome("abaabbaa"))
print(s.longestPalindrome("aacabdkacaa"))

File: leetcode/typescript/5_longest_palindromic_substring.ts
//   b a b a d
// b 1 x x x 1
// a   1 x 1 x
// b.    1 x x
// a.      1 x
// d.        1

function longestPalindrome(s: string): string {
    let res = s[0];
    const n = s.length;
    const dp: boolean[][] = [];
    for (let i = 0; i < n; i++) {
        dp.push(new Array<boolean>(n).fill(false));
        dp[i][i] = true;
    }

    for (let i = n - 2; i >= 0; i--) {
        for (let j = i+1; j < n; j++) {
            if (s[i] != s[j]) continue;

            if (j - i === 1 || dp[i+1][j-1]) {
                dp[i][j] = true;
                if (res.length < j - i + 1) {
                    res = s.slice(i, j+1);
                }
            }
        }
    }
    return res;



};